### PR TITLE
Webpack 3.X support.

### PR DESCRIPTION
**Pull request by MasseGuillaume**

### CODE REVIEW

**Comment by MasseGuillaume** [comment]
@scala-jenkins we want to delete you bro!

**Comment by MasseGuillaume** [comment]
https://travis-ci.org/scalacenter/scalajs-bundler/builds/266312852

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  global:
+    - TRAVIS_NODE_VERSION="8"
+cache:
+  directories:
+    - $HOME/.ivy2/cache
+    - $HOME/.sbt/boot/
+    - $HOME/.coursier/cache
+before_install:
+  # Install Node.js and yarn.
+  - rm -rf ~/.nvm && git clone https://github.com/creationix/nvm.git ~/.nvm && (cd ~/.nvm && git checkout `git describe --abbrev=0 --tags`) && source ~/.nvm/nvm.sh && nvm install $TRAVIS_NODE_VERSION
```
Comment (sjrd) [inline .travis.yml]: Easier:
```
- nvm install $TRAVIS_NODE_VERSION
- nvm use $TRAVIS_NODE_VERSION
```
Done.

See also https://github.com/scala-js/scala-js-env-jsdom-nodejs/blob/v1.0.0-M1/.travis.yml#L17-L22

Comment (MasseGuillaume) [inline .travis.yml]: nice I will update this on as well: https://github.com/scalameta/metadoc/blob/master/.travis.yml#L23

**Comment by MasseGuillaume** [comment]
@sjrd do you think it's possible to have HMR for Scala.js ?

https://webpack.js.org/guides/hot-module-replacement/

**Comment by MasseGuillaume** [comment]
15 mins to run the full test suite. Not bad travis.

**Comment by sjrd** [comment]
> do you think it's possible to have HMR for Scala.js ?

Not in any meaningful way, no. There are discussions about this in existing tickets on scala-js/scala-js.

**Comment by MasseGuillaume** [comment]
https://travis-ci.org/scalacenter/scalajs-bundler/builds/266317103#L7445-L7451

```
[error] (sbt-web-scalajs-bundler/*:scripted) Failed tests:
[error] 	sbt-web-scalajs-bundler / play
[error] (sbt-scalajs-bundler/*:scripted) Failed tests:
[error] 	sbt-scalajs-bundler / sharedconfig
[error] 	sbt-scalajs-bundler / global-namespace
[error] 	sbt-scalajs-bundler / static
[error] 	sbt-scalajs-bundler / global-namespace-with-jsdom-unit-testing
```

**Comment by MasseGuillaume** [comment]
All scripted tests should pass now :dagger: :fireworks:

**Comment by MasseGuillaume** [comment]
Fix #101 Fix #157

File: sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/ScalaJSBundlerPlugin.scala [hunk 2]
```diff
@@ -553,6 +535,8 @@ object ScalaJSBundlerPlugin extends AutoPlugin {
         ReloadWorkflowTasks.webpackTask(fastOptJS)
       }.value,
 
+      installWebpackDevServer := npmUpdate.value,
```
Comment (julienrf) [inline sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/ScalaJSBundlerPlugin.scala]: Do we need to keep this task at all? It is used only to scope the version of `webpack-dev-server` to use, I guess.

File: sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/global-namespace-with-jsdom-unit-testing/common.webpack.config.js [hunk 3]
```diff
@@ -11,14 +11,20 @@ module.exports = {
...
+  config.module.loaders.push(
+    {
+      test: require.resolve(modName),
+      loader: "expose-loader?" + globalModules[modName]
+    }
+  );
+});
+
+
+module.exports = config;
```
Comment (julienrf) [inline sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/global-namespace-with-jsdom-unit-testing/common.webpack.config.js]: Why not keep the immutable style?

File: sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/global-namespace/webpack.config.js [hunk 4]
```diff
@@ -15,11 +15,16 @@ Object.keys(config.entry).forEach(function(key) {
...
-  }
+    use: [{
+      loader: "expose-loader",
+      options: globalModules[pkg]
+    }]
+  });
 });
 
-module.exports = config;
+module.exports = config;
```
Comment (julienrf) [inline sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/global-namespace/webpack.config.js]: Same comment here.

File: sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/static/build.sbt [hunk 5]
```diff
@@ -39,6 +39,12 @@ InputKey[Unit]("html") := {
 
 TaskKey[Unit]("checkSize") := {
   val artifactSize = IO.readBytes((webpack in (Compile, fullOptJS)).value.head).length
-  val expected = 19790
-  assert(artifactSize == expected, s"expected: $expected, got: $artifactSize")
+  val sizeLow = 15000
+  val sizeHigh = 25000
```
Comment (julienrf) [inline sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/static/build.sbt]: How did you choose these numbers?

File: sbt-web-scalajs-bundler/src/sbt-test/sbt-web-scalajs-bundler/play/build.sbt [hunk 6]
```diff
@@ -7,7 +7,8 @@ val client =
       libraryDependencies += "org.scala-js" %%% "scalajs-dom" % "0.9.1",
       npmDependencies in Compile ++= Seq(
         "snabbdom" -> "0.5.3",
-        "font-awesome" -> "4.7.0"
+        "font-awesome" -> "4.7.0",
+        "url-loader" -> "0.5.9"
```
Comment (julienrf) [inline sbt-web-scalajs-bundler/src/sbt-test/sbt-web-scalajs-bundler/play/build.sbt]: Why is this loader needed now?

File: sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/sharedconfig/build.sbt [hunk 7]
```diff
@@ -12,10 +12,11 @@ libraryDependencies += "org.scala-js" %%% "scalajs-dom" % "0.9.1"
...
   "webpack-merge" -> "4.1.0",
-  "file-loader" -> "0.10.1",
-  "image-webpack-loader" -> "3.3.0",
-  "css-loader" -> "0.27.0",
-  "style-loader" -> "0.16.0"
+  "file-loader" -> "0.11.2",
+  "image-webpack-loader" -> "3.3.1",
+  "css-loader" -> "0.28.5",
+  "style-loader" -> "0.18.2",
+  "url-loader" -> "0.5.9"
```
Comment (julienrf) [inline sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/sharedconfig/build.sbt]: Why is this loader needed now?

File: sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/ScalaJSBundlerPlugin.scala (hunk 2 above)
Comment (MasseGuillaume) [inline sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/ScalaJSBundlerPlugin.scala]: Yes and also no to break user's code. Let's keep it until we have a breaking release?

File: sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/static/build.sbt (hunk 5 above)
Comment (MasseGuillaume) [inline sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/static/build.sbt]: Bumping the webpack version changed the artifact size. I did a range check similar to the other test. Should I just put the exact value?

File: sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/global-namespace-with-jsdom-unit-testing/common.webpack.config.js (hunk 3 above)
Comment (MasseGuillaume) [inline sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/global-namespace-with-jsdom-unit-testing/common.webpack.config.js]: It's no the same syntax in v2, V3 or v1:

https://github.com/webpack-contrib/expose-loader#usage

```javascript
// webpack v1 usage
module: {
  loaders: [
    { test: require.resolve("jquery"), loader: "expose-loader?$" }
  ]
}

// webpack v2 usage
module: {
  rules: [{
          test: require.resolve('jquery'),
          use: [{
              loader: 'expose-loader',
              options: '$'
          }]
      }]
}
```

File: sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/global-namespace/webpack.config.js (hunk 4 above)
Comment (MasseGuillaume) [inline sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/global-namespace/webpack.config.js]: ibid. do you want to keep the immutable style with the new syntax?

File: sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/global-namespace-with-jsdom-unit-testing/common.webpack.config.js (hunk 3 above)
Comment (MasseGuillaume) [inline sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/global-namespace-with-jsdom-unit-testing/common.webpack.config.js]: https://webpack.js.org/guides/migrating/#module-loaders-is-now-module-rules

**Comment by MasseGuillaume** [comment]
scroll the logs

```
org.apache.http.conn.HttpHostConnectException: Connect to localhost:7357 [localhost/127.0.0.1, localhost/0:0:0:0:0:0:0:1] failed: Connection refused (Connection refused)
```

grep for 7357

```
# Give the server some time to launch
# It's hacky, but this is just a test..
$ sleep 3000
```

:man_facepalming:

**Comment by easel** [comment]
Looks good to me. I'm generally sanguine about adding a few extra npm dependencies to keep the out-of-the-box experience robust.

File: sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/ScalaJSBundlerPlugin.scala (hunk 2 above)
Comment (julienrf) [inline sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/ScalaJSBundlerPlugin.scala]: Next release will be a breaking release since it upgrades the webpack default version.

File: manual/src/ornate/changelog.md [hunk 8]
```diff
@@ -1,5 +1,13 @@
 # Changelog
 
+## Version 0.8.0-SNAPSHOT
+
+> Not yet released
+
+- [#160](https://github.com/scalacenter/scalajs-bundler/pull/160): 
+  * webpack 3.5.5 and webpack-dev-server 2.1.1 are now the default version
+  * Use the same npm update in webpack-dev-server and bundle
```
Comment (julienrf) [inline manual/src/ornate/changelog.md]: It is not clear to me what does this sentence mean.

File: sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/ScalaJSBundlerPlugin.scala (hunk 2 above)
Comment (MasseGuillaume) [inline sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/ScalaJSBundlerPlugin.scala]: ok. I scope version to startWebpackDevServer then?

File: sbt-scalajs-bundler/src/main/scala/scalajsbundler/WebpackDevServer.scala [hunk 9]
```diff
@@ -47,28 +45,26 @@ private [scalajsbundler] class WebpackDevServer {
...
+      "node_modules/webpack-dev-server/bin/webpack-dev-server.js",
       "--config",
       configPath.absolutePath,
       "--port",
       port.toString
     ) ++ extraArgs
 
+    logger.info("Starting webpack-dev-server");
+    command.foreach(arg => logger.info("  " + arg))
+    logger.info("workDir: " + workDir.toString)
```
Comment (julienrf) [inline sbt-scalajs-bundler/src/main/scala/scalajsbundler/WebpackDevServer.scala]: Do we want to keep all of these logs at the `info` level? I would put the two last lines at the `debug` level.

File: sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/ScalaJSBundlerPlugin.scala [hunk 10]
```diff
@@ -42,6 +42,10 @@ import scalajsbundler._
   *
   * Version of webpack-dev-server to use.
   *
+  * {{{
+  *   version in webpack := "2.7.1"
```
Comment (julienrf) [inline sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/ScalaJSBundlerPlugin.scala]: typo: should be `version in installWebpackDevServer` (if we keep using this key)

File: sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/ScalaJSBundlerPlugin.scala (hunk 2 above)
Comment (julienrf) [inline sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/ScalaJSBundlerPlugin.scala]: Yes, that ok for me.

File: sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/global-namespace-with-jsdom-unit-testing/common.webpack.config.js (hunk 3 above)
Comment (julienrf) [inline sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/global-namespace-with-jsdom-unit-testing/common.webpack.config.js]: I would just use the following:

~~~ javascript
module.export = {
  module: {
    rules: Object.keys(globalModules).map(function (modName) {
      return {
        test: require.resolve(modName),
        use: [{ loader: 'expose-loader', options: globalModules[modName] }]
      }
    })
  }
};
~~~

File: sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/static/build.sbt (hunk 5 above)
Comment (julienrf) [inline sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/static/build.sbt]: This range looks very wide, we should try with just `19000` and `21000`

File: .travis.yml [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+    - $HOME/.coursier/cache
+before_install:
+  # Install Node.js and yarn.
+  - nvm install 6
+  - nvm use 6
+  - node --version
+  - curl -o- -L https://yarnpkg.com/install.sh | bash -s -- --version 0.24.6
+  - export PATH=$HOME/.yarn/bin:$PATH
+script:
+  - sbt scripted
```
Comment (julienrf) [inline .travis.yml]: Could you do a separate PR for this?

**Comment by MasseGuillaume** [comment]
Closing it for now. I will rebase on the recent changes.
